algorithms/src/snark/marlin/ahp/ahp.rs

```diff
@@ -217,29 +245,51 @@ impl<F: PrimeField, MM: MarlinMode> AHPForR1CS<F, MM> {
             .map(|(c, x)| x.iter().zip_eq(&lag_at_beta).map(|(x, l)| *x * l).sum::<F>() * c)
             .sum::<F>();
 
+        let zeta_squared = zeta.square();
+
         #[rustfmt::skip]
         let lincheck_sumcheck = {
-            let mut lincheck_sumcheck = LinearCombination::empty("lincheck_sumcheck");
+            let mut lincheck_sumcheck = LinearCombination::empty("placeholder");
+            // Add rowcheck
+            let mut rowcheck = LinearCombination::empty("lincheck_sumcheck");
+            for (i, combiner) in batch_combiners.iter().enumerate() {
+            rowcheck
+                .add((z_a_s_at_beta[i] * z_b_s_at_beta[i] - z_c_s_at_beta[i]) * combiner, "s_m")
+                .add((z_a_s_at_beta[i] + zeta * z_b_s_at_beta[i] + zeta_squared
+                    * z_c_s_at_beta[i] - f_s_at_beta[i]) * combiner, "s_l");
```

Comment:
I suggest instead of constructing the linear combination for the rowcheck as `(z_a_beta * z_b_beta - z_c_beta) * s_m`, we should construct it as `(z_a * z_b_beta - z_c) * s_m_beta`. This saves  proof size: for a batch size of `n`,we only need to provide 1 evaluation of `s_m_beta` and `n` evaluations of `z_b_beta`, as opposed to `n` evaluations each of `z_a_beta`, `z_b_beta`, and `z_c_beta`. This results in a savings of ~2n field elements.

Similarly for the lookup part, we should do it as `(z_a + zeta * z_b + zeta * z_c) * s_l_beta`.